torchdynamo/variables/tensor.py

```diff
@@ -65,45 +77,101 @@ def visit(n: torch.fx.Node):
 
         return torch.fx.node.map_arg((node.args, node.kwargs), visit)
 
+    @staticmethod
+    def run_proxy(proxy, args, kwargs, nnmodule):
+        op = proxy.node.op
+        if op == "call_function":
+            return proxy.node.target(*args, **kwargs)
+        elif op == "call_method":
+            return getattr(args[0], proxy.node.target)(*args[1:], **kwargs)
+        elif op == "call_module":
+            assert nnmodule is not None
+            return nnmodule(*args, **kwargs)
+        assert False, op
+
+    @classmethod
+    def wrap_to_fake_tensor(cls, e, fake_mode):
+        if type(e) in (torch.Tensor, torch.nn.Parameter):
+            return cls.wrap_fake_exception(lambda: fake_mode.from_tensor(e))
+        else:
+            return e
+
+    @staticmethod
+    def wrap_fake_exception(fn):
+        try:
+            return fn()
+        except UnsupportedFakeTensorException as e:
+            raise FakeTensorError(
+                f"Unsupported: {e.reason} with fake tensor propagation. "
+                "Run with config.fake_tensor_propagation=False"
+            ) from e
+
     @classmethod
     def create(cls, tx, proxy, example_value=None, nnmodule=None, **options):
```

Comment:
We should factor this out into an FX pass that can live in PyTorch fx passes proper